controller/common/manage_children.go

```diff
@@ -126,6 +129,9 @@ func deleteChildren(client *dynamicclientset.ResourceClient, parent *unstructure
 func updateChildren(client *dynamicclientset.ResourceClient, updateStrategy ChildUpdateStrategy, parent *unstructured.Unstructured, observed, desired map[string]*unstructured.Unstructured) error {
 	var errs []error
 	for name, obj := range desired {
+		if parent.GetNamespace() == "" && obj.GetNamespace() != "" {
```

Comment:
I think the desired objects here come directly from the hook response, which means they've never been through the API server to fill in defaults for unset fields.

In that case, an empty namespace could either mean the child is cluster-scoped, or the child is namespaced and the user intended for the namespace to be defaulted to "default". The latter is arguably user error (it would be reasonable to require explicit namespaces on namespaced children of cluster-scoped parents), so my main concern is whether we will end up surfacing an error that tells the user what they did wrong.

Currently, I would guess that the code will attempt a cluster-scoped REST request (since the parent is cluster-scoped)  and I'm worried the error will look like a generic 404 "the server could not find the requested resource", which would leave the user scratching their head.

Would you be willing to take a look at plumbing info about whether a resource is cluster-scoped or namespaced, from discovery down into ResourceClient? Then we could check that here and return a specific error for the case when the user failed to give a namespace for a namespaced child object. It might also make sense to use that discovery info to decide if a parent is namespaced or not, in some places.

Reply:
Sorry it took me so long to get around to this @enisoc . I'be been able to dig into the dynamic client a some more a do some tests. I think you've already plumbed the scope of the resource through. If one tries to create a Namespaced resource without setting the namespace in the object you'll see a log entry similar to:

```
I0728 23:32:39.289957       1 manage_children.go:189] ClusterRole /my-clusterrole: creating RoleBinding /my-clusterrole
E0728 23:32:39.290119       1 controller.go:236] failed to sync cluster-parent "rbac.authorization.k8s.io/v1:ClusterRole::my-clusterrole": can't reconcile children for ClusterRole /my-clusterrole: an empty namespace may not be set during creation
```

Looks like this is coming from [k8s.io/client-go/rest/request.go](https://sourcegraph.com/github.com/kubernetes/client-go@0b97d57a869d33ee2fe3cc567d0f8553c70357b0/-/blob/rest/request.go#L645-651)

Reply:
Thanks for investigating! That error looks pretty clear.